KAFKA-3522: Add RocksDBTimestampedStore

mjsax opened the pull request:
Part of KIP-258. adds only internal classes (we can merge right away).

mjsax committed:
KAFKA-3522: Add RocksDBWithTimestampsStore

mjsax commented:
Call for review @guozhangwang @bbejeck @vvcephei

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacadeTest.java:
```diff
@@ -0,0 +1,257 @@
+import static org.junit.Assert.fail;
+
+@RunWith(EasyMockRunner.class)
+public class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacadeTest {
```
The purpose of this test is, to catch interface changes if we upgrade RocksDB. Using reflections, we make sure the `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacade` maps all methods from `DBOptions` and `ColumnFamilyOptions` to/from `Options` correctly.

vvcephei replied:
very nice!

Actually, can you add that statement as a class javadoc, for posterity?

vvcephei replied:
Was it possible for you to test the test? I.e., how do you know the test does what it's supposed to do?

mjsax replied:
I tested it "manually" by removing one method in `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacade` -- this lets the test fail. Is this sufficient for you?

mjsax replied:
\cc @vvcephei

vvcephei replied:
Yep! Sorry, I lost your reply in the shuffle.

That's sufficient. It looked right to me, just wanted to make sure we've seen it fail at least once.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+    private ColumnFamilyHandle withTimestampColumnFamily;
+
+    // the following option objects will be created in the constructor and closed in the close() method
+    private RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacade userSpecifiedOptions;
```
In contrast to `RocksDBStore`, we use column families. Thus, we cannot pass in `Options` object that is used by users to specify custom RocksDB options via `StreamsConfig`; instead we need to use `DBOptions` and `ColumnFamilyOptions` -- thus, we need to translated from `Options` into those two classed via this helper class.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+            Files.createDirectories(dbDir.getParentFile().toPath());
+
+            try {
+                db = RocksDB.open(dbOptions, dbDir.getAbsolutePath(), columnFamilyDescriptors, columnFamilies);
```
We try to open the DB with both column families -- this might fail if only one exist. For this case, we create the CF and retry to open the DB afterwards

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+        }
+    }
+
+    private byte[] getInternal(final byte[] rawKey) {
```
All operations are defined over both CF; ie, we do dual put/get operation to migrate data from default CF (that does not store any timestamps) to the new CF

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StoreProxyUtils.java:
```diff
@@ -0,0 +1,35 @@
+
+import org.apache.kafka.common.serialization.LongSerializer;
+
+class StoreProxyUtils {
```
This class will get more helper method later...

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+        dbDir = new File(new File(context.stateDir(), parentDir), name);
+
+        try {
+            final List<ColumnFamilyDescriptor> columnFamilyDescriptors = new ArrayList<ColumnFamilyDescriptor>(2) {
```
nit: I personally don't like the double brace initialization for Java collections, but since this subjective feel free to ignore this comment.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+                return rawValueWithUnknownTimestamp;
+            }
+
+            return null;
```
return an empty `Array` instead?

mjsax replied:
`null` indicates not found and should be correct. Let me know if you disagree.

bbejeck replied:
Yeah I agree, forgot about returning `null` from store to indicate not found

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+            if (sstFileNames != null && sstFileNames.length > 0) {
+                try {
+                    db.compactRange(true, 1, 0);
+                    db.compactRange(true, 1, 0);
```
why two calls here to `db.compactRange`? Do we need to make a call for each column family?

mjsax replied:
On oversight from my side. I _think_ we can just call it once (will remove the duplicate line).

The docs are not very specific if this triggers compaction for all CF, but I believe yes. There is also an API to compact one CF, but I don't see any reason to use it. Thoughts?

mjsax replied:
Thinking about this one more, I actually believe that this would only compact default CF -- I'll add this to the `RocksDBAccessor` interface and pass in CF handlers.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+        try (final WriteBatch batch = new WriteBatch()) {
+            for (final KeyValue<byte[], byte[]> record : records) {
+                if (record.value == null) {
+                    batch.delete(noTimestampColumnFamily, record.key);
```
For the operations involving the `noTimestampColumnFamily` maybe we could put a guard condition around each operation `if(noTimesampColumnFamilyNotEmpty) {..}` and the `boolean` flag is set when opening the db and check the size of the default column family with something like `approximateNumEntries` but only checking the `noTimestampColumnFamily`.

mjsax replied:
Addressed this with the new `RocksDBAccessor` interface.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java:
```diff
@@ -0,0 +1,679 @@
+        innerIterWithTimestamp.seekToFirst();
+        final RocksIterator innerIterNoTimestamp = db.newIterator(noTimestampColumnFamily);
+        innerIterNoTimestamp.seekToFirst();
+        final RocksDbIterator rocksDbIterator = new RocksDbIterator(name, innerIterWithTimestamp, innerIterNoTimestamp);
```
Will this work with callers? Will we inspect each value and insert an unknown timestamp for records from default column family?

EDIT: NM I didn't see the `RocksDbIterator` class below when writing this comment.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -82,7 +82,7 @@
     private final String name;
     private final String parentDir;
-    private final Set<KeyValueIterator> openIterators = Collections.synchronizedSet(new HashSet<>());
+    private final Set<KeyValueIterator<Bytes, byte[]>> openIterators = Collections.synchronizedSet(new HashSet<>());
```
This and other stuff in this class is only small code cleanups and reordering of method. There is no actual change.

vvcephei replied:
:+1: Thanks for doing it.

bbejeck reviewed (commented):
Thanks for the hard work @mjsax.  I made an initial pass and left some comments.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -500,7 +472,7 @@ public Bytes peekNextKey() {
     }
 
-    private class RocksDBRangeIterator extends RocksDbIterator {
+    static class RocksDBRangeIterator extends RocksDbIterator {
```
Ok. This is an actual change. Need to make this class `static` to be reused in new `RocksDBTimestampedStore` class. For this reason, we need to pass in `openIterators` in the constructor now.

vvcephei replied:
Makes sense. If it's static, then we could also extract it to a separate file, which might be better than having one of the siblings define a class that both siblings use.

Also, it would reduce the LOC in this class, which is nice for readability.

mjsax replied:
Ack

mjsax commented:
Updated this with `RocksDBAccessor` interface, that allows us to implement single/dual column-family access, depending if there is data in the old column family or not.

bbejeck reviewed (commented):
Thanks for the updates @mjsax.  I've made another pass, and I like the approach of determining when to stop using both column families.

Regarding a test for `RocksDBTimestampedStore` I have a proposal,

1.  Open a plain `RocksDB` give it the two expected column families and populate the `RocksDB.DEFAULT_COLUMN_FAMILY` with a known number of records, then close the plain `RocksDB`.
2. Open an instance of `RocksDBTimestampedStore` then do a series of gets on the known keys and assert all returned with the unknown timestamp, then close.
3. Open the plain `RocksDB again and assert that `RocksDB.DEFAULT_COLUMN_FAMILY` is empty.

WDYT?

mjsax commented:
@bbejeck Added a test as requested :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -546,7 +519,18 @@ public Bytes peekNextKey() {
         @Override
         public void restoreAll(final Collection<KeyValue<byte[], byte[]>> records) {
-            rocksDBStore.restoreAllInternal(records);
+            try (final WriteBatch batch = new WriteBatch()) {
```
`restoreAllInternal()` is only called here, so I think it's better to inline it.

guozhangwang replied:
+1

guozhangwang replied:
Now I see why we need to make `name` package-private now; think it is okay.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java:
```diff
@@ -0,0 +1,619 @@
+
+    @Test
+    public void shouldMigrateDataFromDefaultToTimestampColumnFamily() throws Exception {
+        prepareOldStore();
```
This test let checkstyle fail because method was too long. Extracted some part into `prepareOldStore()` to fix checkstyle.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java:
```diff
@@ -0,0 +1,619 @@
+
+        rocksDBStore.close();
+
+        verifyOldAndNewColumnFamily();
```
Similar as above: extracted into own method to avoid checkstyle issue

bbejeck reviewed (commented):
thanks for the updates and test @mjsax, overall looks good, I'd like to make one more pass

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacade.java:
```diff
@@ -0,0 +1,1363 @@
+ *
+ * This class do the translation between generic {@link Options} into {@link DBOptions} and {@link ColumnFamilyOptions}.
+ */
+class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsFacade extends Options {
```
Neat!

Although, it seems like more of an adapter than a facade ;)

mjsax replied:
Ack. Renaming.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:
```diff
@@ -42,7 +42,7 @@
 import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.addInvocationRateAndCount;
 
-class RocksDBSegmentedBytesStore implements SegmentedBytesStore {
+public class RocksDBSegmentedBytesStore implements SegmentedBytesStore {
```
I couldn't find the reference that makes it necessary to make this public...

mjsax replied:
Ack. Reverting.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,842 @@
+ *
+ * Note that the use of array-typed keys is discouraged because they result in incorrect caching behavior.
+ * If you intend to work on byte arrays as key, for example, you may want to wrap them with the {@code Bytes} class,
+ * i.e. use {@code RocksDBTimestampedStore<Bytes, ...>} rather than {@code RocksDBTimestampedStore<byte[], ...>}.
```
This comment seems misplaced. IIUC, the class already restricts the key type to `Bytes`.

mjsax replied:
c&p issue (from `RocksDBStore`). At some point, the class had generics and it seems while refactoring this we missed to update this JavaDoc. Fixing in both classes.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StoreProxyUtils.java:
```diff
@@ -0,0 +1,35 @@
+            return null;
+        }
+        final byte[] rawValueWithUnknownTimestamp = new byte[8 + rawValue.length];
+        // TODO: should we use `ByteBuffer` instead of `System.arraycopy` ?
```
I don't think it matters. This implementation is pretty straightforward; it doesn't seem like the bytebuffer really has any advantage.

If the others agree, we should get rid of the todo.

mjsax replied:
It the same question as https://github.com/apache/kafka/pull/6151#discussion_r249229217

I would like to get a uniform agreement/policy what to use and will update the code accordingly after we made a decision. \cc @guozhangwang

guozhangwang replied:
I'd say just remove the TODO; using either is fine here.

mjsax replied:
As discussed in person, I update to use `ByteBuffer` what should be our default API to use to modify byte arrays.

vvcephei reviewed (commented):
Thanks for the PR @mjsax .

Overall, it looks good. I left a few comments...

mjsax commented:
Updated this.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
```diff
@@ -0,0 +1,1363 @@
+ *
+ * This class do the translation between generic {@link Options} into {@link DBOptions} and {@link ColumnFamilyOptions}.
+ */
+class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter extends Options {
```
Out of scope of this: we should consider have a KIP in the next release to refactor APIs of config setters as:

```
public interface RocksDBConfigSetter {

    /**
     * Set the rocks db options for the provided storeName.
     * 
     * @param storeName     the name of the store being configured
     * @param options       the Rocks DB options
     * @param configs       the configuration supplied to {@link org.apache.kafka.streams.StreamsConfig}
     */
    void setConfig(final String storeName, final DBOptions dbOptions, final ColumnFamilyOptions cfOptions, final Map<String, Object> configs);
}
```

And then we can choose to call the other constructor of 

```
public Options(final DBOptions dbOptions,
      final ColumnFamilyOptions columnFamilyOptions) {..}
```

mjsax replied:
Sure. Can you please create a Jira for tracking? Otherwise it might get dropped.

guozhangwang replied:
https://issues.apache.org/jira/browse/KAFKA-7869

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java:
```diff
@@ -0,0 +1,1363 @@
+        // doing (3) to (5) below
+        // TODO enabling this breaks StreamThreadTest#shouldRecorverFromInvalidOffsetExceptionOnRestoreAndFinishRestore
+        // -> unclear why
+        //columnFamilyOptions.setMemTableConfig(new VectorMemTableConfig());
```
Looked at https://github.com/facebook/rocksdb/blob/master/options/options.cc#L367 there seems no code for setting memtable config here.

I'd suggest we just mimic the code here instead of following the FAQ since that page may well obsoleted.

mjsax replied:
Ack. Will update the comment accordingly.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    void openDB(final ProcessorContext context) {
```
Meta: this and the existing class still share a lot of common code; I'm wondering if it is possible to just add a flag to the existing class, based on which we can decide whether use Options v.s. DBOptions / CFOptions, and use RocksDBAcccessor.

mjsax replied:
Yes. That works. We can share more code.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+
+        try {
+            final List<ColumnFamilyDescriptor> columnFamilyDescriptors = asList(
+                new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
```
Just following on my other idea about collapsing into a single class here: maybe instead of naming it as keyValueWithTimestamp, we just name it as:

1) "default" -> version 2.1-
2) "2.2" -> version 2.2 to now.

And the flag can just be indicating if it is 1) or 2) above; in the future if we need to do this again we can then have:

1) "default" -> version 2.1-
2) "2.2" -> version 2.2 - 2.5 (just made that up).
3) "3.0" -> version 3.0 - now.

etc.

mjsax replied:
We cannot rename default column family (and we cannot delete it either) -- it's there all the time. I can still rename the new CF to `2.2` if you want. Let me know.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+
+                final RocksIterator noTimestampsIter = db.newIterator(noTimestampColumnFamily);
+                noTimestampsIter.seekToFirst();
+                if (noTimestampsIter.isValid()) {
```
I cannot find in rocksDB docs whether newIterator.seekToFirst().isValid() is a safe / efficient way to check emptiness, could we had a unit test for this specific case?

1) starting the DB with empty cf at the beginning.
2) starting the DB with non-empty cf at the beginning.
3) same as 2), but delete all keys and restart the DB and check again.

mjsax replied:
Isn't this already covered via `RocksDBTimestampedStoreTest#shouldMigrateDataFromDefaultToTimestampColumnFamily`? Or do I miss something?

mjsax replied:
\cc @guozhangwang -- comments?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+                    columnFamilies.get(1).close();
+                    columnFamilies.clear();
+                    db.close();
+                    try {
```
QQ: why we need to close and restart it here? Can we just skip line 202 - 211 ?

mjsax replied:
Good catch!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+
+
+
+    private interface RocksDBAccessor {
```
Again following my meta comment here: we can have one impl of `dbAccessor` here per upgrade version, like 

1) 2.1- only (i.e. only one CF used).
2) 2.2+ only (i.e. two CF but only access the second one).
3) 2.1- -> 2.2+ upgrade (i.e. the dual access one).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+                nextWithTimestamp = getKeyValueWithTimestamp();
+            }
+
+            if (nextNoTimestamp == null) {
```
The logic here is a bit over-complicated to me, can we simplify to sth like standard sort-merge here? We have e.g. `AbstractMergedSortedCacheStoreIterator` for a similar pattern.

mjsax replied:
`AbstractMergedSortedCacheStoreIterator` uses `peekNextKey` that is not available on `RocksDBIterator` -- thus, we need to peek manually -- don't think we can share code here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+                    db.delete(withTimestampColumnFamily, wOptions, rawKey);
+                } catch (final RocksDBException e) {
+                    // String format is happening in wrapping stores. So formatted message is thrown from wrapping stores.
+                    throw new ProcessorStateException("Error while removing key %s from store " + name, e);
```
I'd suggest try-catch each line separately since the underlying `RocksDBException` would not tell you which line actually went wrong, and this piece of info would be very useful for trouble shooting; ditto below.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+
+        @Override
+        public void putAll(final List<KeyValue<Bytes, byte[]>> entries, final WriteBatch batch) throws RocksDBException {
+            for (final KeyValue<Bytes, byte[]> entry : entries) {
```
I think we can still use `write(batch);` here for efficiency?

mjsax replied:
`write(batch)` should be done by the caller -- that we call it in the other accessor is a bug -- will rename the method to `prepareBatch()`.

guozhangwang replied:
Not sure I can follow this: could you elaborate a bit more? I was thinking to use the RocksDB's

```
void write(final WriteOptions writeOpts, final WriteBatch updates)
```

API for the putAll call, is it possible?

mjsax replied:
That is exactly what is use. Compare `RocksDB#putAll()`:
```
    @Override
    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
        try (final WriteBatch batch = new WriteBatch()) {
            dbAccessor.prepareBatch(entries, batch);
            write(batch);
        } catch (final RocksDBException e) {
            throw new ProcessorStateException("Error while batch writing to store " + name, e);
        }
    }
```

It calls `dbAccessor.prepareBatch(entries, batch);` (already renamed from `putAll -> prepareBatch`) that is this method. We do the call "outside" because only who the batch is prepared differs.

Does this make sense

guozhangwang replied:
Yup.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,838 @@
+
+        @Override
+        public long approximateNumEntries() throws RocksDBException {
+            return db.getLongProperty(noTimestampColumnFamily, "rocksdb.estimate-num-keys")
```
Which do you think is more accurate, this or `db.getLongProperty("rocksdb.estimate-num-keys");` directly?

mjsax replied:
If CF is not specified, not all CF all accessed by default CF only -- thus, we need to add both.

mjsax replied:
I extended `RocksDBTimestampedStoreTest` accordingly to test `approximateNumEntries()`, too.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbTimestampedKeyValueBytesStoreSupplier.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.kafka.streams.state.KeyValueBytesStoreSupplier;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+public class RocksDbTimestampedKeyValueBytesStoreSupplier implements KeyValueBytesStoreSupplier {
```
We can probably save this class if we consolidate on one RocksDBStore.

mjsax replied:
Not sure, we still need to pass in the flag into RocksDB in which mode we want to use it.

guozhangwang replied:
UPDATE: can we just modify `RocksDbKeyValueBytesStoreSupplier` to allow passing in a flag, based on which we will construct `RocksDB` v.s. `TRocksDB`?

mjsax replied:
Ack

guozhangwang reviewed (commented):
Made a pass on non-testing code.

mjsax commented:
Updated this. Reworked `RocksDB` and `RocksDBTimestampStore` to share more code. Both use `RocksDBAccessor` now. Also consolidate the two corresponding tests.

Rebased to get BloomFilter changes that got merged recently.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbIterator.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public synchronized KeyValue<Bytes, byte[]> next() {
```
Though this is not introduced in this PR: this function seems not needed?

mjsax replied:
Yes. Maybe we should add the check if the iterator is open? Similar to `hasNext()` ?

mjsax replied:
Never mind. `super.next()` calls `hasNext()` anyway. This part is covered. Removing this method.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -81,17 +78,19 @@
+    final String name;
     private final String parentDir;
-    private final Set<KeyValueIterator> openIterators = Collections.synchronizedSet(new HashSet<>());
+    final Set<KeyValueIterator<Bytes, byte[]>> openIterators = Collections.synchronizedSet(new HashSet<>());
```
Why `openIterators` needs to be package-private now?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -269,68 +247,39 @@ public synchronized void put(final Bytes key,
+    public synchronized byte[] get(final Bytes key) {
+        validateStoreOpen();
+        try {
+            return dbAccessor.get(key.get());
```
For `put / putAll` etc the exception is captured inside the dbAccessor, while for `get` it is captured in the caller, is it intentional?

guozhangwang replied:
If not, we should move the exception capturing logic inside the dbAccessor as well.

mjsax replied:
Fair question. If we move all try-catch into the dbAccessor, we need to duplicate more code. For example, `db.flush()` could throw and catching the exception in `RocksDB#flush()` is a single place -- if we move it into the dbAccessor, we need to to try-catch in both implementations.

Thoughts?

mjsax replied:
`put` is different, because you requested to catch `put()` and `get()` within `DualAccessor` separately -- thus, I moved the code inside.

For `putAll()` does also throw (main pattern is that the caller catches).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -269,68 +247,39 @@ public synchronized void put(final Bytes key,
-        final byte[] value = get(key);
+        final byte[] oldValue;
+        try {
+            oldValue = dbAccessor.getOnly(key.get());
```
Ditto here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -372,16 +317,16 @@ public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
+        final long numEntries;
         try {
-            value = db.getLongProperty("rocksdb.estimate-num-keys");
+            numEntries = dbAccessor.approximateNumEntries();
```
Ditto here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -417,131 +374,176 @@ public synchronized void close() {
+
+        void flush() throws RocksDBException;
+
+        void restoreAllInternal(final Collection<KeyValue<byte[], byte[]>> records,
```
This function seems to be used as the same logic with `prepareBatch` now?

mjsax replied:
The problem is, that one takes `<byte[],byte[]` while the other `<Bytes,byte[]>` we can avoid one only, if we loop to one collection first and create new `KeyValue` with new format. Both solutions are not ideal -- however, I prefer code duplication over runtime overhead to loop through the collection twice.

guozhangwang replied:
Ah I see. I agree that have two functions are preferable.

On the other hand, could we rename this one to `prepareBatchBytes` to indicate the logical consistency?

mjsax replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -417,131 +374,176 @@ public synchronized void close() {
         open = false;
         closeOpenIterators();
-        options.close();
+        dbAccessor.close();
```
For line 376 above, which is out of scope of this PR: thinking about it again now I'm not sure if Streams should be responsible for closing the open iterators when closing the database; this logic is a bit awkward and also subject to errors with concurrent access from IQ. We may instead just educate users to always close the iterators themselves (for both IQ and processor). WDYT?

mjsax replied:
Well. It seems to be a nice "safe guard" -- as you mention out-of scope: maybe create jira and we can discuss there? not sure atm.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
```
This is a meta question for my own understanding: my take is that after https://github.com/apache/kafka/pull/6204 we will let this store to extend `TimestampedBytesStore` as well, right?

mjsax replied:
Correct.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+    private ColumnFamilyHandle noTimestampColumnFamily;
+    private ColumnFamilyHandle withTimestampColumnFamily;
+
+    RocksDBTimestampedStore(final String name) {
```
Are these two constructor necessary?

mjsax replied:
Not really -- it's c&p from `RocksDBStore` that also has both. For both stores, actually only one constructor is used, so we could remove one constructor on both.

vvcephei replied:
:+1: It looks like the two-arg constructor is unused.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+
+
+
+    private class WithTimestampColumnFamilyAccessor extends SingleColumnFamilyAccessor {
```
nit: I'd suggest we remove this (and also the other default db accessor in the other class) class and call `SingleColumnFamilyAccessor(columnFamilies.get(1))`. Reason is that here we make the assumption that `withTimestampColumnFamily` (and `noTimestampColumnFamily` in the other class) is already not-null but that depends on the impl today. This type of style is a bit vulnerable to future bugs that cause NPE.

mjsax replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+        }
+    }
+
+    private class DualColumnFamilyAccessor implements RocksDBAccessor {
```
nit: maybe we can make it just a general accessor that takes two parameters: `oldCF` and `newCF`? Or we can do this generalizing in the future if you'd like to hard-code for now.

mjsax replied:
If we make one, how does it now that it need to upgrade stuff or not? Seems to imply a `null` check for each operation if both or only one CF should be accessed? That would imply runtime overhead. Also, I _think_ the code would be a little bit harder to read. Thoughts?

guozhangwang replied:
We will create the `DualColumnFamilyAccessor` only when at the construction time if we found there are two CFs and the old CF is not empty right?

mjsax replied:
Yes.

guozhangwang replied:
So that means we do not need to do extra check inside this accessor impl right?

vvcephei replied:
Personally, I like the existing design. The `DualColumnFamilyAccessor` has to do a lot of extra checking that isn't necessary if there's just one cf to deal with. If we collapse them into one class with a flag, we pay for it with a lot of branching.

One thing I did find confusing was reasoning about the fact that the Dual accessor is embedded in this (Timestamped) class, and the Single accessor is embedded in the parent (non-timestamped) class. But, we're using it as an accessor for this (the child) class. This seems unnecessarily convoluted, and it's a little hard to see if it's actually ok, or just coincidentally ok, since the parent and child APIs are only semantically, rather than actually, different.

It seems simpler to understand if we pull both accessors out into separate classes that take `db`, `name`, `options`, etc as constructor arguments, rather than closing over protected state.

bbejeck replied:
+1 from me as well for putting both accessors into separate classes

vvcephei replied:
I looked at it closer. I still think it's better to split them out, but I also don't think it's a correctness issue right now, so I'd be fine with merging what you have.

mjsax replied:
Let's to some refactoring as follow ups. It's internal anyway. Would like to keep the PRs focused to get stuff in and merged :)

guozhangwang replied:
Actually I was only suggesting to make the current Dual accessor to be more general
currently it assumes the `old` to be `default`, and `new` to be `withTimestamp`.
what I was suggesting is only to make these two parameterized; so that in the future we only have two accessor impls:

1) XX-CF only; which we already do in this PR.
2) XX-to-YY upgrade: old XX CF to YY CF upgrade accessor.

All that being said, I'm okay with such refactoring as follow-ups.

mjsax replied:
This refactoring is already done: `DualAccessor` has a constructor with two CF parameters.

There is one missing piece: the conversion function is hard coded. After `TimestampedByteStore` PR is merged, I will refactor this to pass in a `RecordConverter` that and will pass in `TimestampedByteStore#convertValueToNewFormat()`

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+                    nextNoTimestamp = null;
+                    iterNoTimestamp.next();
+                } else {
+                    if (comparator.compare(nextNoTimestamp.key.get(), nextWithTimestamp.key.get()) <= 0) {
```
if (comparator.compare(nextNoTimestamp.key.get(), nextWithTimestamp.key.get()) == 0) we need to advance on both ends while only returning the one from with-timestamp iterator, otherwise we may get duplicates returned.

mjsax replied:
This should never happen, because each key should be stored only once. Either in old or new CF. Or do I miss something?

guozhangwang replied:
I think it is still possible. Here's one scenario:

1. last checkpoint at offset 100; all writes goes to old CF.
2. continue writes to old CF til offset 110, but no checkpoint written yet.
3. non-graceful shutdown happens, and upon restarting new CF is used.
4. we start restoring from offset 100 to log-end-offset 110, to the new CF.

Now we ended with data of offsets 100-110 in both CFs.

mjsax replied:
In step (4), we also delete on old CF. Compare https://github.com/apache/kafka/pull/6149/files#diff-be43584f61033a47b5422775f5c5efdaR154

guozhangwang replied:
I see. Do we guarantee that concurrent IQ will not see duplicated results with the db-accessor updating logic as well? If yes, we can save this check.

mjsax replied:
That should be safe. Note, that IQ is "disabled" if we are not in RUNNING state.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+        }
+
+        private KeyValue<Bytes, byte[]> getKeyValueNoTimestamp() {
+            return new KeyValue<>(new Bytes(iterNoTimestamp.key()), getValueWithUnknownTimestamp(iterNoTimestamp.value()));
```
I'd suggest one optimization here, in order to save array-copying every time while we scan over the old CF: we keep the original key and value bytes for `no-timestamp` iterator, and only when we've already decided to assign from this iterator to `next` then we will call `getValueWithUnknownTimestamp` to do the bytes array copying.

mjsax replied:
Ack.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java:
```diff
@@ -0,0 +1,267 @@
+
+
+        // iterating should not migrate any data, but return all key over both CF (plus surrogate timestamps for old CF)
+        final KeyValueIterator<Bytes, byte[]> it = rocksDBStore.all();
```
Could we also test for range(from, to) queries?

mjsax replied:
Ack

guozhangwang reviewed (commented):
Made another pass over the PR.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,411 @@
+                     final ColumnFamilyOptions columnFamilyOptions) {
+        final List<ColumnFamilyDescriptor> columnFamilyDescriptors = asList(
+            new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
+            new ColumnFamilyDescriptor("keyValueWithTimestamp".getBytes(StandardCharsets.UTF_8), columnFamilyOptions));
```
Did @guozhangwang suggest to rename this DF to `2.2`? I actually think the descriptive name might be better. It seems like it'll be less work in the long run to remember what exactly is different about the different CFs.

mjsax replied:
He did suggest but there was no final agreement. I personally don't care too much.

bbejeck replied:
what about naming `keyValueWithTimestamp_2.2`?

guozhangwang replied:
I don't feel strong about my suggestion either :) Will leave it to anyone who has a strong feeling here.

mjsax committed:
Github comments

mjsax committed:
Github comments

vvcephei reviewed (commented):
Thanks, @mjsax !

I'm still looking over it, but I wanted to submit one batch of feedbacks for you.

-John

mjsax committed:
Remove unused constructors

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java:
```diff
@@ -0,0 +1,300 @@
+public class RocksDBTimestampedStoreTest extends RocksDBStoreTest {
+    final String unknownTimestampString = new String(new LongSerializer().serialize(null, ConsumerRecord.NO_TIMESTAMP));
+
+    String getTimestampPrefix() {
```
nit: this is unused

bbejeck reviewed (commented):
Thanks for the update, Overall looks good to me, I want to take another pass over the unit test later this evening.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStoreTest.java:
```diff
@@ -0,0 +1,300 @@
+        // approx: 4 entries on old CF, 1 in new CF
+        assertThat(rocksDBStore.approximateNumEntries(), is(5L));
+
+        // should add new key10 to new CF
```
nit: comment states `key10` added, but actual key put is `key8`

bbejeck reviewed (approved):
One minor nit otherwise LGTM.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -0,0 +1,399 @@
+            }
+
+            if (nextNoTimestamp == null) {
+                if (nextWithTimestamp == null) {
```
should both iterators also be reporting `!isValid` here as well? I'm finding he rocksdb iterator api a little confusing...

I guess if we never allow a null key into the store, then this is an effective way to check for the end of the iteration.

mjsax replied:
`null` keys are not allowed -- that is the assumption in this check. We can add the `!isValid` as safe guard if you want.

vvcephei replied:
I don't feel strongly about it. If we enforce the "no null keys" invariant, then they are equivalent.

It seems mildly confusing that we essentially have two different methods of determining when the iterator has run out of data. I leave it up to you.

mjsax replied:
\cc @guozhangwang WDYT?

mjsax replied:
Ack. Updating this.

vvcephei reviewed (approved):
Hey @mjsax ,

There are a few minor discussions, but overall, it looks good to me.

Thanks!
-John

mjsax committed:
Github comments

guozhangwang commented:
LGTM. @mjsax please feel free to merge after @vvcephei 's comments.

mjsax committed:
Github comments

mjsax commented:
Updated this. Will merge when Jenkins is green.

mjsax commented:
https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/19162/
```
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Aborted due to timeout.
	at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
	at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
	at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:89)
	at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:260)
	at kafka.api.AdminClientIntegrationTest.testElectPreferredLeaders(AdminClientIntegrationTest.scala:1273)
```

Retest this please.
